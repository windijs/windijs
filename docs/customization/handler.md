---
outline: deep
---

<script setup lang="ts">
import Title from "@/Title.vue"
import { ref } from "vue"

const ts = ref(false)
</script>

<Title head="Handler" v-model:ts="ts" />

The handler refers to the function unit that **handles different props or cases**. We can create any utility by combining different handlers, just like building blocks.

**Note**: Before going deep into these handlers, you should read the [Utility Customization Document](/customization/utility) first, make sure that you understand the structure of the API before continuing.

## Built-in Handlers

### configHandler

### colorHandler

### cssHandler

### numberHandler

### pxHandler

### remHandler

### degHandler

### msHandler

### spacingHandler

### fractionHandler

### callHandler

### setupHandler

### genericHandler

## Handler Helpers

### css

### prop

### meld

### guard

### handler

### isHandler

## Custom Handler

You are able to customize handler by extending built-in handler or recreating a new handler.

### Extending built-in handler

You can extract the handler that you feel redundant into another independent handler. For example:

```js
configHandler(colors, "backgroundColor", "--w-bg-opacity")
```

can be extracted into

```js
export function backgroundColorHandler () {
  return configHandler(colors, "backgroundColor", "--w-bg-opacity")
}
```

Another example, `configHandler` doesn't support handling Array by default, `fontFamilyHandler` convert Array to string first, then pass it to `configHandler`.

<template v-if="ts">

```ts
export function fontFamilyHandler<T extends object> (fonts: T) {
  const cssFonts = {} as {[key in keyof typeof fonts]: string};
  for (const [key, value] of Object.entries(fonts)) {
    cssFonts[key as keyof typeof fonts] = Array.isArray(value) ? value.join(",") : value as string;
  }
  return configHandler(cssFonts as T, "fontFamily");
}
```

</template>

<template v-else>

```js
export function fontFamilyHandler (fonts) {
  const cssFonts = {};
  for (const [key, value] of Object.entries(fonts)) {
    cssFonts[key] = Array.isArray(value) ? value.join(",") : value;
  }
  return configHandler(cssFonts, "fontFamily");
}
```

</template>

One more example, when you use `genericHandler`, you usually want to extract it separately.
This example supports `bg[0x1c1c1e], bg[0xff0]`, ....

<template v-if="ts">

```ts
export function backgroundGenericHandler () {
  return genericHandler<{ [key: string]: StyleObject }>("backgroundColor", (prop) => {
    if (isNumber(prop)) {
      return "#" + (+prop).toString(16);
    }
    return prop;
  });
}
```

</template>

<template v-else>

```js
export function backgroundGenericHandler () {
  return genericHandler("backgroundColor", (prop) => {
    if (isNumber(prop)) {
      return "#" + (+prop).toString(16);
    }
    return prop;
  });
}
```

</template>

### Creating new handler

Basically, the built-in handlers is capable for most cases, but sometimes you may want to create your own handler. In this case, you should use [handler](#handler) api.

A handler in fact is a simple object that has 3 elements, `type`, `meta`, `get`. 

- type
    - Explain: The type of the handler.
    - Type: `"number" | "color" | "config" | "css" | "style" | "call" | "spacing" | "fraction" | "generic" | "setup" | "guard" | "meld" | String`
- meta: 
    - Explain: The meta data of the handler.
    - Type: `object | undefined`
- get:
    - Explain: The function that handles props.
    - Type: `(prop: string) => T`

Let's take [fractionHandler](#fractionhandler) as an example.

```js
import { buildStatic, handler, isFraction, fracToPercent } from "windijs" 

export function fractionHandler (propertyOrBuildFunc) {
  // the build func, return StyleObject
  const build = typeof propertyOrBuildFunc === "function" ? propertyOrBuildFunc : value => buildStatic(propertyOrBuildFunc, value);
  return handler("fraction", p => isFraction(p) ? build(fracToPercent(p)) : undefined)
}
```

And if you are using TypeScript, you can also use type assertion.

```ts
import { buildStatic, handler, isFraction, fracToPercent } from "windijs" 
import type { StyleObject, StyleProperties, BuildFunc, Handler } from "windijs"

export function fractionHandler<T extends object = { [key:string]: StyleObject }> (propertyOrBuildFunc: StyleProperties | StyleProperties[] | BuildFunc) {
  // the build func, return StyleObject
  const build: BuildFunc = typeof propertyOrBuildFunc === "function" ? propertyOrBuildFunc : value => buildStatic(propertyOrBuildFunc, value);
  return {
    type: "fraction",
    get: p => isFraction(p) ? build(fracToPercent(p)) : undefined,
  } as Handler<T>;
}
```