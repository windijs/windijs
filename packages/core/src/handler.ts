import { CSSMap, CSSObject, StyleObject, prop } from "@windi/helpers";
import { buildFontSize, buildKeyframes, buildProperty } from "./builder";
import { configHandler, cssHandler, genericHandler, guard } from "./api";
import { prefixAnimation, prefixNotHidden } from "./prefixer";

import { isNumber } from "@windi/shared";

export const animateHandler = <T extends string> (name: T, value: string | CSSObject, keyframes?: Record<string, CSSObject>) => {
  const map: CSSMap = new Map();
  const setValue = (o: CSSObject) => Object.entries(o).map(([k, v]) => map.set(k, v));
  if (keyframes) setValue(buildKeyframes(name, keyframes));
  setValue(typeof value === "string" ? buildProperty([prop`-webkit-animation`, "animation"], value) : prefixAnimation(value));

  return guard(name, cssHandler(map));
};

export function backgroundGenericHandler () {
  return genericHandler<{ [key: string]: StyleObject } & { $: {[key: string]: StyleObject} }>("backgroundColor", (prop) => {
    if (isNumber(prop)) {
      return "#" + (+prop).toString(16);
    }
    return prop;
  });
}

export function divideXReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-divide-x-reverse": "1",
  }));
}

export function divideYReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-divide-y-reverse": "1",
  }));
}

export function fontFamilyHandler<T extends object> (fonts: T) {
  const cssFonts = {} as {[key in keyof typeof fonts]: string};
  for (const [key, value] of Object.entries(fonts)) {
    cssFonts[key as keyof typeof fonts] = Array.isArray(value) ? value.join(",") : value as string;
  }
  return configHandler(cssFonts as T, "fontFamily");
}

export function fontSizeHandler<T extends object> (sizes: T) {
  return configHandler(sizes, value => {
    if (typeof value === "string") return buildFontSize(value);
    if (Array.isArray(value) && value[0]) {
      if (value[1] == null || typeof value[1] === "string") return buildFontSize(value[0], value[1]);
      if (typeof value[1] === "object" && value[1] != null) return buildFontSize(value[0], undefined, value[1]);
    }
  });
}

export function spaceBetweenXReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-space-x-reverse": "1",
  }));
}

export function spaceBetweenYReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-space-y-reverse": "1",
  }));
}
