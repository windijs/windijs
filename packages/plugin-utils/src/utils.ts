import { EntryOptions, PluginEnv, PluginOptions, ResolvedPluginEnv, ResolvedPluginOptions } from "./types";
import { basename, join, resolve } from "path";
import { existsSync, mkdirSync, readFileSync, rmSync, writeFileSync } from "fs";

export const DefaultOptions: ResolvedPluginOptions = {
  include: ["./src/**/*.{vue,jsx,tsx}"],
  exclude: ["**/node_modules"],
  config: {},
  configEntry: "./windi.config",
  env: {
    nodeModulesEntry: "./node_modules",
    globalEntry: "./src/windi-global.d.ts",
    moduleEntry: "./src/windi-module.d.ts",
    variants: {
      lib: "@windijs/variants",
      module: true,
      global: true,
    },
    utilities: {
      lib: "@windijs/utilities",
      module: true,
      global: true,
    },
    config: {
      lib: "windijs",
      module: true,
      global: false,
    },
  },
};

export const isProduction = process.env.NODE_ENV === "production";

export const replaceEntry = (input: string) => input.replace(/@windijs\/(core|colors|helpers|config|shared)/g, "windijs");

export const filterConflict = (src: string, entries: string[]) => entries.filter(i => !src.match(new RegExp(`(const|let|var)\\s+${i}\\s*=`)));

export const injectImports = (code: string, imports: Record<string, string[]>) => Object.entries(imports).map(([k, v]) => `import { ${v.join(", ")} } from '${k}';\n`).join("") + code;

export const injectConfig = (code: string, path: string) => `import windiUserConfig from '${path}';\n` + code;

export const injectHelper = (code: string, helper: string, pkg: string) => code.includes(helper) ? code : injectImports(code, { [isProduction ? pkg : "windijs"]: [helper] });

export const requireImports = (code: string, imports: Record<string, string[]>) => code.replace("var ", Object.entries(imports).map(([k, v]) => `var { ${v.join(", ")} } = require('${k}');\n`).join("") + "var ");

export const requireConfig = (code: string, path: string) => code.replace("var ", `var windiUserConfig = require('${path}');\nvar `);

export const requireHelper = (code: string, helper: string, pkg: string) => code.includes(helper) ? code : requireImports(code, { [isProduction ? pkg : "windijs"]: [helper] });

export function injectStyleLoader (code: string) {
  code = injectHelper(code, "useStyleLoader", "@windijs/helpers");
  const start = code.match(/(const|let|var|function|export)\s+/)?.index ?? 0;
  code = code.slice(0, start) + "useStyleLoader(windiUserConfig.styleLoader);\n" + code.slice(start);
  return code;
}

export const declModule = (name: string, contents: string | string[]) => `declare module '${name}' {\n` + (Array.isArray(contents) ? contents.map(i => "  " + i + ";\n").join("") : contents) + "}\n";

export const updateEnv = (entry: string | false, content: string | undefined) => {
  if (!entry) return;
  if (content) writeFileSync(entry, "// This file was auto generated by windijs, you can put it into your `.gitignore`, please do not edit it directly.\n" + content);
  else if (existsSync(entry)) rmSync(entry);
};

export function readPackage (dir: string) {
  return JSON.parse(readFileSync(join(dir, "package.json")).toString()) as { types?: string, module?: string, main?: string, exports?: Record<string, string | { require?: string, import?: string, types?: string }> };
}

export function readModule (dir: string, module?: string) {
  const pkg = readPackage(dir);
  const mod = module ?? pkg.module;
  if (!mod) throw new Error(`Couldn't find module entry for '${basename(dir)}'`);
  if (!pkg.types) throw new Error(`Counln't find types tentry for '${basename(dir)}'`);
  const dts = readFileSync(join(dir, pkg.types)).toString();
  const mjs = readFileSync(join(dir, mod)).toString();
  return { pkg, dts, mjs };
}

export function resolveEntry (options: EntryOptions, defaultValue: Required<EntryOptions>): Required<EntryOptions> {
  const resolvedEntry = defaultValue;
  if (options.global != null) resolvedEntry.global = options.global;
  if (options.lib != null) resolvedEntry.lib = options.lib;
  if (options.module != null) resolvedEntry.module = options.module;
  return resolvedEntry;
}

export function resolveEnv (env?: PluginEnv): ResolvedPluginEnv {
  const resolvedEnv = DefaultOptions.env;
  if (!env) return resolvedEnv;
  for (const [k, v] of Object.entries(env)) {
    if (typeof v !== "object") {
      // @ts-ignore
      resolvedEnv[k] = v;
    }
  }
  if (env.config) resolvedEnv.config = resolveEntry(env.config, resolvedEnv.config);
  if (env.utilities) resolvedEnv.utilities = resolveEntry(env.utilities, resolvedEnv.utilities);
  if (env.variants) resolvedEnv.variants = resolveEntry(env.variants, resolvedEnv.variants);

  return resolvedEnv;
}

export function resolveOptions (options: PluginOptions): ResolvedPluginOptions {
  const resolvedOptions: ResolvedPluginOptions = { ...DefaultOptions, ...options, env: resolveEnv(options.env) };
  // convert relative path to absolute path
  resolvedOptions.configEntry = resolve(resolvedOptions.configEntry);
  resolvedOptions.include = resolvedOptions.include.map(i => resolve(i));
  resolvedOptions.exclude = resolvedOptions.exclude.map(i => resolve(i));
  return resolvedOptions;
}

export function refreshDir (dir: string) {
  if (existsSync(dir)) rmSync(dir, { recursive: true });
  mkdirSync(dir);
  return dir;
}
