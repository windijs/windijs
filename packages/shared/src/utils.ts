import type { FractionObject, Negative, RangeTuple } from "./types";

/**
 * Check if value is a number
 * @param value {string}
 * @returns {boolean}
 */
export function isNumber (value: string) {
  // @ts-ignore
  return !isNaN(value);
}

export function isFraction (amount: string): boolean {
  return /^\d+\/\d+$/.test(amount);
}

export function isSize (amount: string): boolean {
  return /^-?(\d+(\.\d+)?)+(rem|em|px|rpx|vh|vw|ch|ex|cm|mm|in|pt|pc)$/.test(amount);
}

export function roundUp (num: number, precision = 0): number {
  precision = Math.pow(10, precision);
  return Math.round(num * precision) / precision;
}

export function fracToPercent (amount: string): string | undefined {
  const matches = amount.match(/[^/]+/g);
  if (!matches || matches.length < 2) return;
  const a = +matches[0];
  const b = +matches[1];
  return roundUp((a / b) * 100, 6) + "%";
}

export const hasKey = <T extends object>(obj: T, k: keyof any): k is keyof T => k in obj;

export function commaJoin (...items: (string|number|undefined)[]) {
  return items.filter(i => i).join(", ");
}

export function parenWrap (key: string, value: string) {
  return key + "(" + value + ")";
}

export function camelToDash (str: string) {
  return str.replace(/([A-Z])/g, v => `-${v.toLowerCase()}`);
}

export function dashToCamel (str: string) {
  return str.replace(/(-[a-z])/g, v => v.toUpperCase().replace("-", ""));
}

export function hash (str: string): string {
  str = str.replace(/\r/g, "");
  let hash = 5381;
  let i = str.length;

  while (i--) hash = ((hash << 5) - hash) ^ str.charCodeAt(i);
  return (hash >>> 0).toString(36);
}

export function indent (value: string, count: number = 2) {
  return " ".repeat(count) + value;
}

export const entries = <T extends object | Map<string, any>> (t: T) => t instanceof Map ? t.entries() : Object.entries(t);

export function negative <T extends Record<string | number, string>> (t: T): Negative<T> {
  return Object.fromEntries(Object.entries(t).filter(([, v]) => !/^0(px|rem|em)?$/.test(v)).map(([k, v]) => ["-" + k, "-" + v])) as Negative<T>;
}

export function range<S extends number, E extends number> (start: S, end: E): RangeTuple<S, E> {
  return new Array(end - start).fill(0).map((_, i) => i + start) as RangeTuple<S, E>;
}

export function fractions<S extends number, E extends number> (start: S, end: E) {
  return Object.fromEntries((range(start, end) as number[]).map(i => range(1, i as 0).map(v => [v + "/" + i, v / i * 100] as [string, number]).map(([k, v]) => [k, v % 1 === 0 ? v + "%" : v.toFixed(6) + "%"])).flat()) as FractionObject<S, E>;
}

export function spacings<T extends Array<any>> (numbers: T) {
  return Object.fromEntries(numbers.map(i => [i, i / 4 + (i === 0 ? "px" : "rem")])) as Record<T[number], string>;
}

export function degrees<T extends number> (numbers: T[]) {
  return Object.fromEntries(numbers.map(i => [i, i + "deg"])) as Record<T, string>;
};

export function scales<T extends number> (numbers: T[]) {
  return Object.fromEntries(numbers.map(i => [i, i / 100 + ""])) as Record<T, string>;
};

export function omit<T extends object, K extends object> (t: T, keys: K) {
  return Object.fromEntries(Object.entries(t).filter(([k]) => !(k in keys))) as Omit<T, keyof K>;
}
