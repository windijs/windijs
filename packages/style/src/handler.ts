import { $in, $var, BuildFunc, CSSDecls, GeneralCSSData, Q, StyleHandler, StyleProxy, abs, attr, blur, brightness, calc, ch, circle, clamp, cm, conicGradient, contrast, counter, counters, css, cubicBezier, deg, dpcm, dpi, dppx, dropShadow, ellipse, em, env, ex, fitContent, fr, getMeta, grad, grayscale, hsl, hsla, hueRotate, hwb, inset, invert, linearGradient, matrix, matrix3d, max, min, minmax, mm, ms, opacity, path, pc, percent, perspective, polygon, pt, pushMetaProp, px, quote, rad, radialGradient, rem, repeat, repeatingConicGradient, repeatingLinearGradient, repeatingRadialGradient, rgb, rgba, rotate, rotate3d, rotateX, rotateY, rotateZ, s, saturate, scale, scale3d, scaleX, scaleY, scaleZ, sepia, sign, skew, skewX, skewY, steps, translate, translate3d, translateX, translateY, translateZ, turn, url, useProxy, vh, vmax, vmin, vw, x } from "@windijs/helpers";

import { handleConfig } from "@windijs/core";
import { hasKey } from "@windijs/shared";

const funcs = { quote, attr, url, $var, path, hwb, dropShadow, counters, circle, ellipse, inset, polygon, matrix, matrix3d, perspective, rotate, rotate3d, rotateX, rotateY, rotateZ, scale, scale3d, scaleX, scaleY, scaleZ, skew, skewX, skewY, translate, translate3d, translateX, translateY, translateZ, steps, calc, clamp, max, min, abs, sign, blur, brightness, contrast, grayscale, invert, opacity, saturate, sepia, rgb, rgba, hsl, hsla, counter, env, minmax, repeat, hueRotate, fitContent, cubicBezier, linearGradient, radialGradient, conicGradient, repeatingConicGradient, repeatingLinearGradient, repeatingRadialGradient };
const units = { percent, deg, grad, rad, turn, s, ms, fr, $in, dpi, dpcm, dppx, x, px, pc, pt, cm, mm, Q, ch, ex, em, rem, vw, vh, vmax, vmin };

export function stylePropertyHandler<T extends Partial<Record<keyof CSSDecls, unknown | BuildFunc<string>>> = {}> (cfg: T = {} as T) {
  const build = (prop: string, value: string) => {
    const meta = getMeta();
    if (meta.props.length === 1) meta.props.push(value);
    return css({ [prop]: value }, undefined, meta);
  };
  return {
    type: "style",
    get: (prop) => useProxy(value => {
      pushMetaProp(prop);
      if (hasKey(cfg, prop)) {
        const cv = cfg[prop];
        const result = typeof cv === "function" ? cv(value) : handleConfig(v => build(prop, v as string), cv as unknown as object, "css", value);
        if (result) return result;
      }
      if (value === "var") return (name: string, defaultValue?: string | undefined) => build(prop, funcs.$var(name, defaultValue));
      if (value in funcs) return (...args: any[]) => build(prop, ((funcs as unknown as { [key: string]: Function })[value])(...args));
      if (value === "in") return useProxy(v => build(prop, units.$in[+v].toString()));
      if (value in units && value !== "color") return useProxy(v => build(prop, (units as any)[value][v].toString()));
      return build(prop, value);
    }),
  } as StyleHandler<GeneralCSSData & StyleProxy<T>>;
}
