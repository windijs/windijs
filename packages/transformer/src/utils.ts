import { Bundle, ScriptTarget, SourceFile, TransformerFactory, createPrinter, createSourceFile, transform } from "typescript";

/**
 * Quickly use typescript transformer to convert the source code.
 * @param src Souce Code of TypeScript
 * @param transformers Typescript Transformers
 * @returns Transformed Code
 */
export function useTransformer (src: string, ...transformers: (TransformerFactory<SourceFile>)[]): string;
export function useTransformer (src: string, ...transformers: (TransformerFactory<SourceFile | Bundle>)[]): string;
export function useTransformer (src: string, ...transformers: (TransformerFactory<any>)[]): string {
  const printer = createPrinter();
  const source = createSourceFile("useTransform.ts", src, ScriptTarget.ESNext, true);
  const result = transform(source, transformers);
  const transformedSourceFile = result.transformed[0] as SourceFile;
  return printer.printFile(transformedSourceFile);
}
