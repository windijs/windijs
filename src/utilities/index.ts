export * from "./base";
export * from "./builder";
export * from "./style";
export * from "./api";
export * from "./bg";
export * from "./font";
export * from "./spacing";
