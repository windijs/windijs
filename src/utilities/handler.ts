import * as funcs from "helpers/funcs";
import * as units from "helpers/unit";

import type { CSSDecls, CSSMap, CSSObject, GeneralCSSData, Handler, StyleObject, StyleProxy } from "types";
import { buildFontSize, buildKeyframes, buildProperty } from "./builder";
import { configHandler, cssHandler, genericHandler, guard, handleConfig } from "./api";
import { css, useProxy } from "helpers";
import { getMeta, pushMetaProp } from "helpers/meta";
import { hasKey, isNumber } from "utils";
import { prefixAnimation, prefixNotHidden } from "./prefixer";

import { prop } from "helpers/common";

export const animateHandler = <T extends string> (name: T, value: string | CSSObject, keyframes?: Record<string, CSSObject>) => {
  const map: CSSMap = new Map();
  const setValue = (o: CSSObject) => Object.entries(o).map(([k, v]) => map.set(k, v));
  if (keyframes) setValue(buildKeyframes(name, keyframes));
  setValue(typeof value === "string" ? buildProperty([prop`-webkit-animation`, "animation"], value) : prefixAnimation(value));

  return guard(name, cssHandler(map));
};

export function backgroundGenericHandler () {
  return genericHandler<{ [key: string]: StyleObject } & { $: {[key: string]: StyleObject} }>("backgroundColor", (prop) => {
    if (isNumber(prop)) {
      return "#" + (+prop).toString(16);
    }
    return prop;
  });
}

export function divideXReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-divide-x-reverse": "1",
  }));
}

export function divideYReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-divide-y-reverse": "1",
  }));
}

export function fontFamilyHandler<T extends object> (fonts: T) {
  const cssFonts = {} as {[key in keyof typeof fonts]: string};
  for (const [key, value] of Object.entries(fonts)) {
    cssFonts[key as keyof typeof fonts] = Array.isArray(value) ? value.join(",") : value as string;
  }
  return configHandler(cssFonts as T, "fontFamily");
}

export function fontSizeHandler<T extends object> (sizes: T) {
  return configHandler(sizes, value => {
    if (typeof value === "string") return buildFontSize(value);
    if (Array.isArray(value) && value[0]) {
      if (value[1] == null || typeof value[1] === "string") return buildFontSize(value[0], value[1]);
      if (typeof value[1] === "object" && value[1] != null) return buildFontSize(value[0], undefined, value[1]);
    }
  });
}

export function spaceBetweenXReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-space-x-reverse": "1",
  }));
}

export function spaceBetweenYReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-space-y-reverse": "1",
  }));
}

type BuildFunc = (value: string) => StyleObject | undefined;

export function stylePropertyHandler<T extends Partial<Record<keyof CSSDecls, unknown | BuildFunc>> = {}> (cfg: T = {} as T) {
  const build = (prop: string, value: string) => {
    const meta = getMeta();
    if (meta.props.length === 1) meta.props.push(value);
    return css({ [prop]: value }, undefined, meta);
  };
  return (prop => useProxy(value => {
    pushMetaProp(prop);
    if (hasKey(cfg, prop)) {
      const cv = cfg[prop];
      const result = typeof cv === "function" ? cv(value) : handleConfig(v => build(prop, v as string), cv as object, "css", value);
      if (result) return result;
    }
    if (value === "var") return (name: string, defaultValue?: string | undefined) => build(prop, funcs.$var(name, defaultValue));
    if (value in funcs) return (...args: any[]) => build(prop, ((funcs as unknown as { [key: string]: Function })[value])(...args));
    if (value === "in") return useProxy(v => build(prop, units.$in[+v].toString()));
    if (value in units && value !== "color") return useProxy(v => build(prop, (units as any)[value][v].toString()));
    return build(prop, value);
  })) as Handler<GeneralCSSData & StyleProxy<T>>;
}
