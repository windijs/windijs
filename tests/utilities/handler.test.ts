import { CSSObject, StyleObject, backgroundGeneric, borderColor, borderStyle, borderStyleConfig, borderWidth, borderWidthConfig, colors, createUtility, use } from "index";
import { guard, meld, useGenericHandler } from "utilities/handler";

import { isNumber } from "utils";

test("useGeneric With Trigger", () => {
  const backgroundGeneric = useGenericHandler<"trigger", { [key: string]: StyleObject }>("trigger", prop => {
    const build = (value: string) => ({
      backgroundColor: value,
    } as CSSObject);
    if (isNumber(prop)) {
      return build("#" + (+prop).toString(16));
    }
    return build(prop);
  });

  const bg = createUtility("bg")
    .use(backgroundGeneric())
    .init();

  expect(bg.trigger["rgb(22, 22, 22)"].css).toMatchSnapshot();
});

test("useGeneric With Trigger2", () => {
  const backgroundGeneric = useGenericHandler<{ [key: string]: StyleObject }>(prop => {
    const build = (value: string) => ({
      backgroundColor: value,
    } as CSSObject);
    if (isNumber(prop)) {
      return build("#" + (+prop).toString(16));
    }
    return build(prop);
  });

  const bg = createUtility("bg")
    .use(guard("test", guard("trigger", backgroundGeneric())))
    .init();

  expect(bg.test.trigger["rgb(22, 22, 22)"].css).toMatchSnapshot();
});

test("guard with meld", () => {
  const border = createUtility("border")
    .use(guard("colors", guard("dark", borderColor(colors))))
    .use(guard("width", meld(borderWidth(borderWidthConfig), borderColor(colors))))
    .init();

  expect(border.colors.dark.amber[100].css).toMatchSnapshot();
  expect(border.width[2].css).toMatchSnapshot();
  expect(border.width.blue).toBeDefined();
  expect(border.width.blue[400].css).toMatchSnapshot();
});

test("two guard with same name should not work", () => {
  const border = createUtility("border")
    .use(guard("width", borderWidth(borderWidthConfig)))
    .use(guard("width", borderColor(colors)))
    .init();

  expect(border.width[2]).toBeDefined();
  expect(border.width.blue).toBeUndefined();
});

test("use single plugin", () => {
  const bg = use("bg", backgroundGeneric());

  expect(bg["rgb(22, 22, 22)"].css).toMatchSnapshot();
  expect(bg.red.css).toMatchSnapshot();
  expect(bg.aliceBlue.css).toMatchSnapshot();
  expect(bg[0xffffff].css).toMatchSnapshot();
});

test("use multi plugin with meld and use", () => {
  const border = use("border", meld(borderWidth(borderWidthConfig), borderColor(colors)));
  const border2 = use("border", meld(borderWidth(borderWidthConfig), borderColor(colors), borderStyle(borderStyleConfig)));

  expect(border[0].css).toMatchSnapshot();
  expect(border.emerald[500].css).toMatchSnapshot();
  expect(border2[4].css).toEqual(border[4].css);
  expect(border2.emerald[500].css).toEqual(border.emerald[500].css);
  expect(border2.dashed.css).toMatchSnapshot();
});
