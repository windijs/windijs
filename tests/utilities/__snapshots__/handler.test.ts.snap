// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`guard with meld 1`] = `
Object {
  "--w-border-opacity": "1",
  "borderColor": "rgba(254, 243, 199, var(--w-border-opacity))",
}
`;

exports[`guard with meld 2`] = `
Object {
  "borderWidth": "2px",
}
`;

exports[`guard with meld 3`] = `
Object {
  "--w-border-opacity": "1",
  "borderColor": "rgba(96, 165, 250, var(--w-border-opacity))",
}
`;

exports[`use multi plugin with meld and use 1`] = `
Object {
  "borderWidth": "0px",
}
`;

exports[`use multi plugin with meld and use 2`] = `
Object {
  "--w-border-opacity": "1",
  "borderColor": "rgba(16, 185, 129, var(--w-border-opacity))",
}
`;

exports[`use multi plugin with meld and use 3`] = `
Object {
  "borderStyle": "dashed",
}
`;

exports[`use single plugin 1`] = `
Object {
  "backgroundColor": "rgb(22, 22, 22)",
}
`;

exports[`use single plugin 2`] = `
Object {
  "backgroundColor": "red",
}
`;

exports[`use single plugin 3`] = `
Object {
  "backgroundColor": "aliceBlue",
}
`;

exports[`use single plugin 4`] = `
Object {
  "backgroundColor": "#ffffff",
}
`;

exports[`useGeneric With Trigger 1`] = `
Object {
  "backgroundColor": "rgb(22, 22, 22)",
}
`;

exports[`useGeneric With Trigger2 1`] = `
Object {
  "backgroundColor": "rgb(22, 22, 22)",
}
`;
